data-processing/common/commands/expand_macros.py

```diff
@@ -0,0 +1,191 @@
+    tex_file: CompiledTexFile
+
+
+@dataclass
+class ExpansionResult:
+    expansions: List[Expansion]
+    latexml_stdout: bytes
+
+
+class ExpandMathMacros(ArxivBatchCommand[ExpansionTask, ExpansionResult]):
+    @staticmethod
+    def get_name() -> str:
+        return "expand-math-macros"
+
+    @staticmethod
+    def get_description() -> str:
+        return "Expand macros and commands used in TeX math environments."
+
+    def get_arxiv_ids_dirkey(self) -> str:
+        return "expanded-sources"
+
+    def load(self) -> Iterator[ExpansionTask]:
+        for arxiv_id in self.arxiv_ids:
+            sources_dir = directories.arxiv_subdir("expanded-sources", arxiv_id)
+            if not os.path.exists(sources_dir):
+                logging.warning(  # pylint: disable=logging-not-lazy
+                    "No directory of TeX sources could be found for paper %s. The macros "
+                    + "in this paper will not be expanded.",
+                    arxiv_id,
+                )
+                continue
+
+            output_dir = directories.arxiv_subdir("normalized-sources", arxiv_id)
+            if os.path.exists(output_dir):
+                logging.warning(
+                    "Directory for macro expansions already exists in %s. Deleting.",
+                    output_dir,
+                )
+                shutil.rmtree(output_dir)
+
+            compiled_tex_dir = directories.arxiv_subdir("compiled-sources", arxiv_id)
+            compiled_tex_files = get_compiled_tex_files(compiled_tex_dir)
+            if len(compiled_tex_files) == 0:
+                logging.warning(  # pylint: disable=logging-not-lazy
+                    "Could not find any TeX files that were successfully compiled for paper %s. "
+                    + "The pipeline will not be able to process this paper further.",
+                    arxiv_id,
+                )
+                continue
+
+            for tex_file in compiled_tex_files:
+                yield ExpansionTask(arxiv_id, output_dir, tex_file)
+
+    def process(self, item: ExpansionTask) -> Iterator[ExpansionResult]:
+
+        # Make a copy of the sources directory that in which the macros will be expanded.
+        shutil.copytree(
+            directories.arxiv_subdir("expanded-sources", item.arxiv_id),
+            item.output_sources_dir,
+        )
+
+        # Save paths to the TeX source file in the new directory.
+        sources_dir = item.output_sources_dir
+        tex_path = os.path.join(sources_dir, item.tex_file.path)
+        absolute_tex_path = os.path.realpath(os.path.abspath(tex_path))
+
+        # Remove '.tex' extension from the name of the TeX file.
+        tex_name = re.sub(r"\.tex$", "", item.tex_file.path)
+
+        latexml_bin = os.path.join(
+            os.path.abspath(os.path.join("perl", "latexml", "bin", "latexml"))
+        )
+        if not os.path.exists(latexml_bin):
+            logging.warning(  # pylint: disable=logging-not-lazy
```

Comment:
Naively I'd assume this should cause the pipeline to fail if the box it's running on isn't setup correctly? If someone didn't want to run this step they could pull it out of the pipeline themselves right?

Reply:
Yes, this is a good point. I will add an exception in the style of those you and Chris have been adding lately.